x-pack/plugins/lens/public/datatable_visualization/transpose_helpers.ts

```diff
@@ -0,0 +1,237 @@
+        metricsColumnArgs
+      );
+    });
+}
+
+function transposeRows(
+  firstTable: Datatable,
+  bucketsColumnArgs: Array<ColumnState & { type: 'lens_datatable_column' }>,
+  formatters: Record<string, FieldFormat>,
+  transposedColumnFormatter: FieldFormat,
+  transposedColumnId: string,
+  metricsColumnArgs: Array<ColumnState & { type: 'lens_datatable_column' }>
+) {
+  const rowsByBucketColumns: Record<string, DatatableRow[]> = groupRowsByBucketColumns(
+    firstTable,
+    bucketsColumnArgs,
+    formatters
+  );
+  firstTable.rows = mergeRowGroups(
+    rowsByBucketColumns,
+    bucketsColumnArgs,
+    transposedColumnFormatter,
+    transposedColumnId,
+    metricsColumnArgs
+  );
+}
+
+/**
+ * Updates column args by adding bucket column args first, then adding transposed metric columns
+ * grouped by unique value
+ */
+function updateColumnArgs(
+  args: Args,
+  bucketsColumnArgs: Array<ColumnState & { type: 'lens_datatable_column' }>,
+  transposedColumnGroups: Array<Array<ColumnState & { type: 'lens_datatable_column' }>>
+) {
+  args.columns = [...bucketsColumnArgs];
+  // add first column from each group, then add second column for each group, ...
+  transposedColumnGroups[0].forEach((_, index) => {
+    transposedColumnGroups.forEach((transposedColumnGroup) => {
+      args.columns.push(transposedColumnGroup[index]);
+    });
+  });
+}
+
+/**
+ * Finds all unique values in a column in order of first occurence
+ * @param table Table to search through
+ * @param formatter formatter for the column
+ * @param columnId column
+ */
+function getUniqueValues(table: Datatable, formatter: FieldFormat, columnId: string) {
+  const values = new Map<string, unknown>();
+  table.rows.forEach((row) => {
+    const rawValue = row[columnId];
+    values.set(formatter.convert(row[columnId]), rawValue);
+  });
+  const uniqueValues = [...values.keys()];
+  const uniqueRawValues = [...values.values()];
+  return { uniqueValues, uniqueRawValues };
+}
+
+/**
+ * Calculate transposed column objects of the datatable object and puts them into the datatable.
+ * Returns args for additional columns grouped by metric
+ * @param metricColumns
+ * @param firstTable
+ * @param uniqueValues
+ */
+function transposeColumns(
+  args: Args,
+  bucketsColumnArgs: Array<ColumnState & { type: 'lens_datatable_column' }>,
+  metricColumns: Array<ColumnState & { type: 'lens_datatable_column' }>,
+  firstTable: Datatable,
+  uniqueValues: string[],
+  uniqueRawValues: unknown[],
+  transposingDatatableColumn: DatatableColumn
+) {
+  const columnGroups = metricColumns.map((metricColumn) => {
+    const originalDatatableColumn = firstTable.columns.find((c) => c.id === metricColumn.columnId)!;
+    const datatableColumns = uniqueValues.map((uniqueValue) => {
+      return {
+        ...originalDatatableColumn,
+        id: getTransposeId(uniqueValue, metricColumn.columnId),
+        name: `${uniqueValue} ${TRANSPOSE_VISUAL_SEPARATOR} ${originalDatatableColumn.name}`,
```

Comment:
I found that the column naming here was not meeting my expectations in two different directions.

1. For simple tables, I wanted to see a table with the unique value only. The metric name is not needed if there's only one metric, as long as the chart title conveys the meaning.

2. I also wanted to see tables with the operation name, like `Operation name > Unique value > Metric name`. This would add context.

For an extreme example of this, try creating a single-row chart with multiple split columns. You'll see how the column labels get confusing:

<img width="957" alt="Screen Shot 2021-03-10 at 6 49 20 PM" src="https://user-images.githubusercontent.com/666475/110714526-8daceb80-81d1-11eb-93fc-f5c12aff6a4b.png">